// angular2localization/src/pipes/locale-number.pipe.js
/**
 * ANGULAR 2 LOCALIZATION
 * An Angular 2 library to translate messages, dates and numbers.
 * Written by Roberto Simonetti.
 * MIT license.
 * https://github.com/robisim74/angular2localization
 */
System.register(['@angular/core', '@angular/common/src/facade/intl', '../services/locale-number', '../services/locale.service'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __extends = (this && this.__extends) || function (d, b) {
        for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var core_1, intl_1, locale_number_1, locale_service_1;
    var LocaleDecimalPipe, LocalePercentPipe, LocaleCurrencyPipe;
    return {
        setters:[
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (intl_1_1) {
                intl_1 = intl_1_1;
            },
            function (locale_number_1_1) {
                locale_number_1 = locale_number_1_1;
            },
            function (locale_service_1_1) {
                locale_service_1 = locale_service_1_1;
            }],
        execute: function() {
            /**
             * 'localedecimal' pipe function.
             */
            LocaleDecimalPipe = (function (_super) {
                __extends(LocaleDecimalPipe, _super);
                function LocaleDecimalPipe(locale) {
                    _super.call(this);
                    this.locale = locale;
                }
                /**
                 * LocaleDecimalPipe transform method.
                 *
                 * @param value The number to be localized
                 * @param defaultLocale The default locale
                 * @param digits The format of the number
                 * @return The locale decimal
                 */
                LocaleDecimalPipe.prototype.transform = function (value, defaultLocale, digits) {
                    if (digits === void 0) { digits = null; }
                    return locale_number_1.LocaleNumber.format(defaultLocale, value, intl_1.NumberFormatStyle.Decimal, digits);
                };
                LocaleDecimalPipe = __decorate([
                    core_1.Pipe({
                        name: 'localedecimal',
                        pure: true
                    }),
                    core_1.Injectable(), 
                    __metadata('design:paramtypes', [locale_service_1.LocaleService])
                ], LocaleDecimalPipe);
                return LocaleDecimalPipe;
            }(locale_number_1.LocaleNumber));
            exports_1("LocaleDecimalPipe", LocaleDecimalPipe);
            /**
             * 'localepercent' pipe function.
             */
            LocalePercentPipe = (function (_super) {
                __extends(LocalePercentPipe, _super);
                function LocalePercentPipe(locale) {
                    _super.call(this);
                    this.locale = locale;
                }
                /**
                 * LocalePercentPipe transform method.
                 *
                 * @param value The number to be localized
                 * @param defaultLocale The default locale
                 * @param digits The format of the number
                 * @return The locale percent
                 */
                LocalePercentPipe.prototype.transform = function (value, defaultLocale, digits) {
                    if (digits === void 0) { digits = null; }
                    return locale_number_1.LocaleNumber.format(defaultLocale, value, intl_1.NumberFormatStyle.Percent, digits);
                };
                LocalePercentPipe = __decorate([
                    core_1.Pipe({
                        name: 'localepercent',
                        pure: true
                    }),
                    core_1.Injectable(), 
                    __metadata('design:paramtypes', [locale_service_1.LocaleService])
                ], LocalePercentPipe);
                return LocalePercentPipe;
            }(locale_number_1.LocaleNumber));
            exports_1("LocalePercentPipe", LocalePercentPipe);
            /**
             * 'localecurrency' pipe function.
             */
            LocaleCurrencyPipe = (function (_super) {
                __extends(LocaleCurrencyPipe, _super);
                function LocaleCurrencyPipe(locale) {
                    _super.call(this);
                    this.locale = locale;
                }
                /**
                 * LocaleCurrencyPipe transform method.
                 *
                 * @param value The number to be localized
                 * @param defaultLocale The default locale
                 * @param currency The current currency
                 * @param symbolDisplay Indicates whether to use the currency symbol
                 * @param digits The format of the number
                 * @return The locale currency
                 */
                LocaleCurrencyPipe.prototype.transform = function (value, defaultLocale, currency, symbolDisplay, digits) {
                    if (symbolDisplay === void 0) { symbolDisplay = false; }
                    if (digits === void 0) { digits = null; }
                    return locale_number_1.LocaleNumber.format(defaultLocale, value, intl_1.NumberFormatStyle.Currency, digits, currency, symbolDisplay);
                };
                LocaleCurrencyPipe = __decorate([
                    core_1.Pipe({
                        name: 'localecurrency',
                        pure: true
                    }),
                    core_1.Injectable(), 
                    __metadata('design:paramtypes', [locale_service_1.LocaleService])
                ], LocaleCurrencyPipe);
                return LocaleCurrencyPipe;
            }(locale_number_1.LocaleNumber));
            exports_1("LocaleCurrencyPipe", LocaleCurrencyPipe);
        }
    }
});
